File: src/journal.ts
import { appendFileSync, existsSync, mkdirSync, readFileSync } from 'node:fs';
import { dirname } from 'node:path';

export interface JournalEntry {
  ts: string;
  kind: string;
  [key: string]: unknown;
}

export function appendJournal(path: string, entry: Omit<JournalEntry, 'ts'>): JournalEntry {
  const full: JournalEntry = { ts: new Date().toISOString(), ...entry } as JournalEntry;
  mkdirSync(dirname(path), { recursive: true });
  appendFileSync(path, JSON.stringify(full) + '\n');
  return full;
}

/** Read all journal entries from a JSONL file. Returns [] if the file doesn't exist. */
export function readJournal(path: string): JournalEntry[] {
  if (!existsSync(path)) return [];
  const content = readFileSync(path, 'utf8').trim();
  if (!content) return [];
  return content.split('\n').map((line) => JSON.parse(line) as JournalEntry);
}

File: python/steward/journal.py
"""Append-only JSONL decision journal.

Every non-trivial decision is recorded here by the executor. The LLM
process never writes the journal — this preserves integrity if a
planner goes off the rails.
"""
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.") + \
        f"{datetime.now(timezone.utc).microsecond // 1000:03d}Z"


def append_journal(path: str | Path, entry: dict[str, Any]) -> dict[str, Any]:
    """Append an entry as a JSONL line, stamping ts. Returns the full entry."""
    full: dict[str, Any] = {"ts": _now_iso(), **entry}
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    with p.open("a", encoding="utf-8") as f:
        f.write(json.dumps(full) + "\n")
    return full


def read_journal(path: str | Path) -> list[dict[str, Any]]:
    """Read all journal entries. Returns [] if the file doesn't exist or is empty."""
    p = Path(path)
    if not p.exists():
        return []
    content = p.read_text(encoding="utf-8").strip()
    if not content:
        return []
    return [json.loads(line) for line in content.split("\n")]
